File: backend/src/main/java/com/ooas/catalog/CatalogService.java
package com.ooas.catalog;

import com.ooas.catalog.dto.InventoryRequest;
import com.ooas.catalog.dto.InventoryResponse;
import com.ooas.catalog.dto.SiteRequest;
import com.ooas.catalog.dto.SiteResponse;
import com.ooas.catalog.dto.SkuRequest;
import com.ooas.catalog.dto.SkuResponse;
import com.ooas.common.ApiException;
import com.ooas.domain.Site;
import com.ooas.domain.SiteInventory;
import com.ooas.domain.Sku;
import com.ooas.repository.SiteInventoryRepository;
import com.ooas.repository.SiteRepository;
import com.ooas.repository.SkuRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

@Service
public class CatalogService {

    private final SkuRepository skuRepository;
    private final SiteRepository siteRepository;
    private final SiteInventoryRepository inventoryRepository;

    public CatalogService(SkuRepository skuRepository, SiteRepository siteRepository, SiteInventoryRepository inventoryRepository) {
        this.skuRepository = skuRepository;
        this.siteRepository = siteRepository;
        this.inventoryRepository = inventoryRepository;
    }

    @Transactional(readOnly = true)
    public List<SkuResponse> listSkus(String search) {
        return skuRepository.search(normalize(search)).stream().map(SkuResponse::from).toList();
    }

    @Transactional
    public SkuResponse createSku(SkuRequest request) {
        if (skuRepository.existsByCodeIgnoreCase(request.code())) {
            throw ApiException.conflict("Ma SKU da ton tai");
        }
        Sku sku = new Sku(request.code().trim().toUpperCase(), request.name().trim(), request.unit().trim(), request.description());
        return SkuResponse.from(skuRepository.save(sku));
    }

    @Transactional
    public SkuResponse updateSku(String id, SkuRequest request) {
        Sku sku = requireSku(id);
        skuRepository.findByCodeIgnoreCase(request.code())
                .filter(existing -> !existing.getId().equals(id))
                .ifPresent(existing -> {
                    throw ApiException.conflict("Ma SKU da ton tai");
                });
        sku.setCode(request.code().trim().toUpperCase());
        sku.setName(request.name().trim());
        sku.setUnit(request.unit().trim());
        sku.setDescription(request.description());
        return SkuResponse.from(sku);
    }

    @Transactional(readOnly = true)
    public List<SiteResponse> listSites(Boolean active, String search) {
        return siteRepository.search(active, normalize(search)).stream()
                .map(site -> SiteResponse.from(site, inventoryRepository.countBySiteId(site.getId())))
                .toList();
    }

    @Transactional
    public SiteResponse createSite(SiteRequest request) {
        if (siteRepository.existsByCodeIgnoreCase(request.code())) {
            throw ApiException.conflict("Ma site da ton tai");
        }
        Site site = new Site(
                request.code().trim().toUpperCase(),
                request.name().trim(),
                request.country().trim(),
                request.seaLeadTime(),
                request.airLeadTime(),
                request.active() == null || request.active()
        );
        siteRepository.save(site);
        return SiteResponse.from(site, 0);
    }

    @Transactional
    public SiteResponse updateSite(String id, SiteRequest request) {
        Site site = requireSite(id);
        siteRepository.findByCodeIgnoreCase(request.code())
                .filter(existing -> !existing.getId().equals(id))
                .ifPresent(existing -> {
                    throw ApiException.conflict("Ma site da ton tai");
                });
        site.setCode(request.code().trim().toUpperCase());
        site.setName(request.name().trim());
        site.setCountry(request.country().trim());
        site.setSeaLeadTime(request.seaLeadTime());
        site.setAirLeadTime(request.airLeadTime());
        if (request.active() != null) {
            site.setActive(request.active());
        }
        return SiteResponse.from(site, inventoryRepository.countBySiteId(site.getId()));
    }

    @Transactional(readOnly = true)
    public List<InventoryResponse> getSiteInventory(String siteId) {
        requireSite(siteId);
        return inventoryRepository.findBySiteIdOrderBySkuCodeAsc(siteId).stream()
                .map(InventoryResponse::from)
                .toList();
    }

    @Transactional
    public InventoryResponse upsertInventory(String siteId, InventoryRequest request) {
        Site site = requireSite(siteId);
        Sku sku = requireSku(request.skuId());
        SiteInventory inventory = inventoryRepository.findBySiteIdAndSkuId(siteId, request.skuId())
                .orElseGet(() -> new SiteInventory(site, sku, 0));
        inventory.setQuantity(request.quantity());
        inventoryRepository.save(inventory);
        return InventoryResponse.from(inventory);
    }

    public Sku requireSku(String id) {
        return skuRepository.findById(id).orElseThrow(() -> ApiException.notFound("Khong tim thay SKU"));
    }

    public Site requireSite(String id) {
        return siteRepository.findById(id).orElseThrow(() -> ApiException.notFound("Khong tim thay site"));
    }

    private String normalize(String value) {
        return StringUtils.hasText(value) ? value.trim() : null;
    }
}

File: frontend/src/services/catalogService.ts
import axiosInstance from '../config/axios';

export interface SkuResponse {
  id: string;
  code: string;
  name: string;
  description?: string;
  unit: string;
  category: string;
  active: boolean;
}

export interface SiteResponse {
  id: string;
  code: string;
  name: string;
  country: string;
  seaLeadTime: number;
  airLeadTime: number;
  active: boolean;
  skuCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface SiteRequest {
  code: string;
  name: string;
  country: string;
  seaLeadTime: number;
  airLeadTime: number;
  active: boolean;
}

export interface InventoryResponse {
  id: string;
  siteId: string;
  siteCode: string;
  siteName: string;
  skuId: string;
  skuCode: string;
  skuName: string;
  unit: string;
  quantity: number;
  updatedAt: string;
}

export const catalogService = {
  listSkus: async (search?: string): Promise<SkuResponse[]> => {
    const params = new URLSearchParams();
    if (search) params.append('search', search);
    const response = await axiosInstance.get(`/api/skus?${params.toString()}`);
    return response.data;
  },
  listSites: async (active?: boolean, search?: string): Promise<SiteResponse[]> => {
    const params = new URLSearchParams();
    if (active !== undefined) params.append('active', String(active));
    if (search) params.append('search', search);
    const response = await axiosInstance.get(`/api/sites?${params.toString()}`);
    return response.data;
  },
  createSite: async (data: SiteRequest): Promise<SiteResponse> => {
    const response = await axiosInstance.post('/api/sites', data);
    return response.data;
  },
  updateSite: async (id: string, data: SiteRequest): Promise<SiteResponse> => {
    const response = await axiosInstance.put(`/api/sites/${id}`, data);
    return response.data;
  },
  getSiteInventory: async (siteId: string): Promise<InventoryResponse[]> => {
    const response = await axiosInstance.get(`/api/sites/${siteId}/inventory`);
    return response.data;
  },
  upsertInventory: async (siteId: string, skuId: string, quantity: number): Promise<InventoryResponse> => {
    const response = await axiosInstance.put(`/api/sites/${siteId}/inventory`, { skuId, quantity });
    return response.data;
  },
};
